### PR TITLE
base: rewrite Monoid module docs

### PR DESCRIPTION
**Pull request by gilligan**
This updates the docs for Monoid analogous to the way i updated the description for Semigroup in #176

### CODE REVIEW

**Comment by Daniel-Diaz** [comment]
The module description explains `<>` and `mempty`, but then all the examples use `mconcat` instead, a function that has not been introduced yet at that point. For someone reading this for the first time, it can be a little confusing.

**Comment by gilligan** [comment]
@Daniel-Diaz i see. Hm hm hm, ok let me try something.. hang on

File: libraries/base/Data/Monoid.hs [hunk 1]
```diff
@@ -16,8 +16,29 @@
...
--- A class for monoids (types with an associative binary operation that
--- has an identity) with various general-purpose instances.
+-- A type @a@ is a 'Monoid' if it provides an associative function ('<>')
+-- that lets you combine any two values of type @a@ into one, and a neutral
+-- element (`mempty`) such that 
+--
+-- > a <> mempty == mempty a == a
+--
+-- ==== __Examples__
+-- >>> Sum 1 <> Sum 2 <> Sum 3 <> Sum 4 :: Sum Int
```
Comment (gilligan) [inline libraries/base/Data/Monoid.hs]: @Daniel-Diaz how about now? I start of both scenarios (Sum and Product) with the invoking `<>` first, then use `mconcat`. I believe this should make it easier to deduce what is going on even without having previously seen `mconcat`.

File: libraries/base/Data/Monoid.hs [hunk 2]
```diff
@@ -16,8 +16,29 @@
 -- Stability   :  experimental
 -- Portability :  portable
 --
--- A class for monoids (types with an associative binary operation that
--- has an identity) with various general-purpose instances.
+-- A type @a@ is a 'Monoid' if it provides an associative function ('<>')
+-- that lets you combine any two values of type @a@ into one, and a neutral
+-- element (`mempty`) such that 
+--
+-- > a <> mempty == mempty a == a
```
Comment (jez) [inline libraries/base/Data/Monoid.hs]: I think you are missing a `<>`?

File: libraries/base/Data/Monoid.hs [hunk 3]
```diff
@@ -16,8 +16,29 @@
...
--- A class for monoids (types with an associative binary operation that
--- has an identity) with various general-purpose instances.
+-- A type @a@ is a 'Monoid' if it provides an associative function ('<>')
+-- that lets you combine any two values of type @a@ into one, and a neutral
+-- element (`mempty`) such that 
+--
+-- > a <> mempty == mempty <> a == a
+--
+-- ==== __Examples__
+-- >>> Sum 1 <> Sum 2 <> Sum 3 <> Sum 4 :: Sum Int
```
Comment (bgamari) [inline libraries/base/Data/Monoid.hs]: I wonder if we should add a sentence briefly introducing `Sum` so it is clear to a new-comer what this example means. For instance,

```
Take, for instance, the example of the 'Sum' monoid, defined by the numerical 
addition operator and the neutral element `0`:
>>> mempty :: Sum Int
Sum 0
>>> Sum 1 <> Sum 2 <> Sum 3 <> Sum 4 :: Sum Int
Sum {getSum = 10}
```
and so on.

File: libraries/base/Data/Monoid.hs [hunk 4]
```diff
@@ -16,8 +16,29 @@
...
+-- Sum {getSum = 10}
+-- >>> mconcat [1,2,3,4] :: Sum Int
+-- Sum {getSum = 10}
+-- >>> mconcat [] :: Sum Int
+-- Sum {getSum = 0}
+-- >>> Product 1 <> Product 2 <> Product 3 <> Product 4 :: Product Int
```
Comment (bgamari) [inline libraries/base/Data/Monoid.hs]: Similarly, we might also consider a brief sentence introducing `Product`.

File: libraries/base/Data/Monoid.hs (hunk 3 above)
Comment (gilligan) [reply libraries/base/Data/Monoid.hs]: done, PTAL

File: libraries/base/Data/Monoid.hs (hunk 4 above)
Comment (gilligan) [reply libraries/base/Data/Monoid.hs]: done, PTAL

File: libraries/base/Data/Monoid.hs [hunk 5]
```diff
@@ -16,8 +16,42 @@
...
+-- The 'Sum' monoid is defined by the numerical addition operator and `0` as neutral element:
+--
+-- >>> mempty :: Sum Int
+-- Sum 0
+-- >>> Sum 1 <> Sum 2 <> Sum 3 <> Sum 4 :: Sum Int
+-- Sum {getSum = 10}
+--
+-- We can combine multiple values in a list into a single value using the `mconcat` function. 
+-- Note that we have to specify the type here since Int has more than one monoid instance:
+--
```
Comment (treeowl) [inline libraries/base/Data/Monoid.hs]: Should be `'Int'`. I wouldn't say it *has" more than one "monoid instance". Perhaps "supports"?

Comment (treeowl) [reply libraries/base/Data/Monoid.hs]: Better, perhaps: is a monoid under several different operations.

File: libraries/base/Data/Monoid.hs [hunk 6]
```diff
@@ -16,8 +16,42 @@
...
+-- >>> mconcat [1,2,3,4] :: Sum Int
+-- Sum {getSum = 10}
+-- >>> mconcat [] :: Sum Int
+-- Sum {getSum = 0}
+--
+-- Another valid monoid instance of Int is 'Product' It is defined by multiplication
+-- and `1` as neutral element:
```
Comment (treeowl) [inline libraries/base/Data/Monoid.hs]: `'Int'` or `@Int@`. Never just `Int`.

File: libraries/base/Data/Monoid.hs [hunk 7]
```diff
@@ -16,8 +16,42 @@
...
+-- >>> Product 1 <> Product 2 <> Product 3 <> Product 4 :: Product Int
+-- Product {getProduct = 24}
+-- >>> mconcat [1,2,3,4] :: Product Int
+-- Product {getProduct = 24}
+-- >>> mconcat [] :: Product Int
+-- Product {getProduct = 1}
+--
+-- ==== __Theoretical Context__
+-- A 'Monoid' is a 'Semigroup' with the added requirement of a neutral element.
+-- Thus any 'Monoid' is a 'Semigroup', but not the other way around.
```
Comment (treeowl) [inline libraries/base/Data/Monoid.hs]: This is not "theoretical context". It looks like mostly a restatement of what's up top. Remove it and integrate whatever you like into the beginning.

File: libraries/base/Data/Monoid.hs (hunk 6 above)
Comment (gilligan) [reply libraries/base/Data/Monoid.hs]: fixed

File: libraries/base/Data/Monoid.hs (hunk 7 above)
Comment (gilligan) [reply libraries/base/Data/Monoid.hs]: fair enough ;) I also just did the same for the semigroup PR..

File: libraries/base/Data/Monoid.hs (hunk 5 above)
Comment (gilligan) [reply libraries/base/Data/Monoid.hs]: That does indeed sound much better. Fixed!

**Comment by gilligan** [comment]
@treeowl all problems fixed? Anything else you’d like to see changed before you’d be willing to ack the changes?

**Comment by gilligan** [comment]
@treeowl gentle ping 😅

**Comment by gilligan** [comment]
Any chance of getting this merged to master?

**Comment by bgamari** [comment]
Merged. Thanks @gilligan!

**Pull request closed without merging**